### ISSUE TITLE
Tracking issue for CompositionEvent support

### ISSUE THREAD

**Issue by murarth** [rust-windowing/winit#1497]
This issue tracks the implementation of IME composition events on each platform, an API initially proposed in #1293. PRs implementing this event should be made against the `composition-event` branch, which will be merged into `master` once all implementations are complete.

* [x] Public API implemented (#1404)
* [ ] Implemented on all platforms
  * [ ] Windows
  * [ ] Mac OS
  * [ ] iOS
  * [x] Linux X11 (#1404)
  * [ ] Linux Wayland
  * [ ] Web
* [ ] Document `CompositionEvent`
* [ ] Add an entry to `CHANGELOG.md`

**Comment by chrisduerr** [comment]
As a consumer of winit I'd recommend against naming these events `CompositionEvent`. Applications that deal with display servers will already know that term in connection with window compositing and it could easily be confused with something related to that.

It also does not fall in line with the existing winit naming of functions like `set_ime_position`.

I think it would be much better to have a more self descriptive name for this event that cannot be confused with anything else related to the winit project. I'd strongly suggest that `IME` should at least be part of the name.

**Comment by pickfire** [comment]
Oh, I thought `CompositionEvent` means some kind of compose key thing, I never used those and don't even know how to use compose key, I had only used `IME` to input characters not in the keyboard.

**Comment by garasubo** [comment]
@pickfire This event naming comes from web specification (Ref: https://developer.mozilla.org/en-US/docs/Web/API/Element/compositionstart_event). And IME will create those composition events.
But as @chrisduerr is pointed out, I agree that calling those events as "composition" is confusing.

**Comment by garasubo** [comment]
Proposed the rename these events to `IMEPreeditEvent` in #1622

**Comment by kaiwk** [comment]
For mac, maybe we can use this API? [composedString](https://developer.apple.com/documentation/objectivec/nsobject/1385416-composedstring?language=objc)

**Comment by kchibisov** [comment]
I've decided to take a look on the proposed events wrt compositing, since I've been working on IME handling in winit for Wayland recently.

Right now I can see that the current API that is being proposed looks more like

```rust
pub enum CompositionEvent {
    CompositionStart(String),
    CompositionUpdate(String, usize),
    CompositionEnd(String),
}
```

If I were modelling it from Wayland side it'll be more like.
```rust
pub enum CompositionEvent {
    /// IME got enabled on the surface you're on.
    CompositionEnabled
    /// We start a composition, a cursors are offsets into string, which could represent a 'highlight' string. `None` for both means hidden.
    CompositionPreedit(text: String, cursor_start: Option<usize>, cursor_end: Option<usize>)
    /// Deletes surrounding text before and after current cursor.
    CompositionDeleteSurroundingText(before_length: usize, after_length: usize)
    /// Composition is done, and that's the text that should be inserted into widget.
    CompositionCommit(text: String)
    /// Composition is done, user must process all events and apply them atomically
    CompositionDone
    /// IME focus got lost, no need to process any IME stuff anymore.
    CompositionDisabled
}
```

So those are just plain events that we have straight from a Wayland protocol. The thing is that it works in some kind of atomic commits way, if you think about it. So on `CompositionDone` you must perform the following:

1. Replace existing preedit string with the cursor.
2. Delete requested surrounding text.
3. Insert commit string with the cursor at its end.
4. Calculate surrounding text to send.
5. Insert new preedit text in cursor position.
6. Place cursor inside preedit text.

Now the question is how to model that with the proposed IME events.

So first of all I don't need `CompostionDone` event, since this is an event when I'd send things downstream, meaning that I don't need `CompositionDeleteSurroundingText`. And it seems like I don't need `DeleteSurroundingText`, since I can model it via `CompositionPreedit`. `CompositionCommit(text: String)` is indeed needed so the user will know when to insert text.

So if we now apply that to the proposed composition event it'll look more like.

```rust
pub enum CompositionEvent {
    CompositionEnabled,
    CompositionPreedit(String, Option<usize>, Option<usize>),
    CompositionCommit(String),
    CompositionDisabled,
}
```

I guess it should make it a bit clear, so we removed 1 event for users to carry about when editing takes place.

However what wasn't mentioned is that how user interacts with IME via requests. Right now
in winit we only have `fn set_ime_position`, which is just some basic stuff. However user may want to
opt for IME handling, via e.g. `fn enable_ime`, so they will react that way on `CompositionEnabled`(basically means that user got IME opened). And they could want to hide IME sometimes, when they interact with a keyboard, but it doesn't require IME things, like a game, where you have a chat with IME support, so they will need `fn disable_ime`.

Also, the user is likely want to set text around the cursor and a cursor position, so they will need `set_surrounding_text`, and also a content type of the current text being edited, with `set_content_type`, so IME can provide better suggestions.

In the end `Window` will get

```rust
fn set_ime_surrounding_text(text: String, cursor: usize);
fn enable_ime_handling();
fn disable_ime_handling();
fn set_contents_type(type: ContentType);
```

The changes shouldn't affect existing work on IME in winit much, and purely about ergonomics and additions.

@garasubo since you've started initial work on updating IME handling in winit, I'm curious whether proposed changes here make sense, and are possible on X11.

@kaiwk As the one who send [composition event implementation](https://github.com/rust-windowing/winit/pull/1669). Are proposed changes sound reasonable to you and possible on macOS, or macOS could need a bit more things to work, nicely?

@chrisduerr As a winit consumer that deals with text input, and that should implement IME handling, does the proposed API sounds ergonomic to implement? One benefit is that you'll know precisely when you actually have IME, so you won't call to `set_ime_position` until you've got an event from a winit that IME was enabled for some window.

Also, since winit doesn't have IME handling on a lot of platforms to land IME handling APIs we can only implement it as a concept on platforms that support it, like X11/macOS, and soon™ Wayland. Platforms that don't support IME in particular could just send IME text via `ReceivedChar` event like IME is working right now.

**Comment by pickfire** [comment]
> One benefit is that you'll know precisely when you actually have IME, so you won't call to set_ime_position until you've got an event from a winit that IME was enabled for some window.

Based on my experience with xinput on X11, I believe it is good to be able to call `set_ime_position` when the IME was enabled rather than updating it for every key press or some weird solution to make some timer to update it once it a while.

**Comment by garasubo** [comment]
> However what wasn't mentioned is that how user interacts with IME via requests.

That's a great idea. But for me, adding those APIs sounds out of scope of this issue. My original motivation is to get IME related event from the window system [to implement CompositionEvent API in the web browser engine](https://github.com/servo/servo/issues/24724). So, I didn't add APIs to control IME from the application.
I think it would be better to have another issue to add those APIs to make this problem simple. I can support x11 implementation.

**Comment by kaiwk** [comment]
> Are proposed changes sound reasonable to you and possible on macOS, or macOS could need a bit more things to work, nicely?

Actually, i'm not very familiar about IME, the PR just fix my own problem, but i'll see what i can do for it. Is there any detailed doc about `surrounding text` and `content type` here?  thanks.

**Comment by kchibisov** [comment]
>So, I didn't add APIs to control IME from the application.
I think it would be better to have another issue to add those APIs to make this problem simple. I can support x11 implementation.

My motivation was to change enum to those 4 variants, so it's in scope of that issue, since I don't quit like the 3 variants we have right now. We can add window methods later however I'd really like to see enum reworked, so users stop calling IME functions without IME being enabled.

**Comment by pickfire** [comment]
Can it commit partially?

**Comment by garasubo** [comment]
> > So, I didn't add APIs to control IME from the application.
> > I think it would be better to have another issue to add those APIs to make this problem simple. I can support x11 implementation.
> 
> My motivation was to change enum to those 4 variants, so it's in scope of that issue, since I don't quit like the 3 variants we have right now. We can add window methods later however I'd really like to see enum reworked, so users stop calling IME functions without IME being enabled.

I see. I meant that adding API to control IME from window (e.g. enable_ime_handling) could be an independent issue, but either way is fine. I can help x11 implementation anyway.

I'm not sure what is the `surrounding text` concept. I think in x11 there is no such concept. Could you explain this more specifically? Maybe part of preedit string to convert, like I type "きょうもいいてんきですね" in IME and try to convert "きょう" to "今日" firstly, then, "きょう" is `surrounding text` and preedit string is "きょうもいいてんきですね"?
Also, when can the second variant of CompositPreedit (cursor_start) be None? In x11, a cursor pos is always defined.

**Comment by kchibisov** [comment]
>I see. I meant that adding API to control IME from window (e.g. enable_ime_handling) could be an independent issue, but either way is fine. I can help x11 implementation anyway.

Yeah, I agree, we can do it separately.

>I'm not sure what is the surrounding text concept. I think in x11 there is no such concept. Could you explain this more specifically? Maybe part of preedit string to convert, like I type "きょうもいいてんきですね" in IME and try to convert "きょう" to "今日" firstly, then, "きょう" is surrounding text and preedit string is "きょうもいいてんきですね"?

It's for when you already have some text around and start editing, so you tell IME that you have certain things around, and your cursor is at some position. You're not required to send this request on Wayland and if you don't know what is around your cursor you don't send it. So if before starting edition you had `きょう` and a cursor right behind that text, you send this string and a cursor position right after the last char, so IME will know that it already has `きょう`, so suggestions are not empty right away, but include your text.

>Also, when can the second variant of CompositPreedit (cursor_start) be None? In x11, a cursor pos is always defined.

On Wayland it's `None` when you should hide a cursor, I'm not sure how to expose that, since Wayland passes negative numbers here, so for rust I've translated that part into `None`. If you have an idea how to translate that concept it'll be nice.

**Comment by kchibisov** [comment]
>Can it commit partially?

The purpose of `Commit` on IMEs is that you insert text to a widget and move cursor to a new position. You can't commit partially, since it's basically 'apply change'. While you're editing you're getting preedits, which informs about the current string, but once user hit a key that commits its input like `Enter`(this is handled by IME) you get commit and start new editing.

**Comment by kchibisov** [comment]
>Actually, i'm not very familiar about IME, the PR just fix my own problem, but i'll see what i can do for it. Is there any detailed doc about surrounding text and content type here? thanks.

Those were just suggestions from Wayland IME protocol that thought should apply to other display servers, since those seems essential for IME.

You can read more about concepts I'm talking about https://github.com/wayland-project/wayland-protocols/blob/master/unstable/text-input/text-input-unstable-v3.xml . If macOS doesn't have those and have something different feel free to post what it has, since I don't really know anything about IME on macOS.

**Comment by chrisduerr** [comment]
> @chrisduerr As a winit consumer that deals with text input, and that should implement IME handling, does the proposed API sounds ergonomic to implement? One benefit is that you'll know precisely when you actually have IME, so you won't call to set_ime_position until you've got an event from a winit that IME was enabled for some window.

I'd like to point out that this could be a potential pitfall? Is the IME popup only going to be visible once I set its position? Or is it always going to pop up and then move around potentially?

I'm also not quite sure what `CompositionPreedit` is for or why it is needed in general. Which of course is something that should be very clear for an API that downstream has to implement.

**Comment by kchibisov** [comment]
I>'d like to point out that this could be a potential pitfall? Is the IME popup only going to be visible once I set its position? Or is it always going to pop up and then move around potentially?

The goal is the following, you get `CompositionEnabled` event from winit, and then you enable IME editing on a window specific window you've got such event. After that all calls to `set_ime_position`, `set_surrounding_text` will work, if you don't call enable nothing will work and IME won't be handled at all, and window won't ever popup. It could be done in a way that calling to `set_ime_position` will result in automatically enabling ime handling, but I'm not sure about it. The goal is to minimize downstream work, since right now they call `set_ime_position` every time without IME being presented.

>I'm also not quite sure what CompositionPreedit is for or why it is needed in general. Which of course is something that should be very clear for an API that downstream has to implement.

It's telling you about string that is currently being edited, but you can't insert it into widget **yet**. This is a quite common name  and concept in IME world tbh. The goal is to display that string to end user, so they know what they're typing, since IME window isn't guaranteed to be presented unlike preedit string, and it makes typing more natural. You generally display that string **inline** in your text editor and user can move a cursor inside it and edit edit while typing.

**Comment by chrisduerr** [comment]
> The goal is the following, you get CompositionEnabled event from winit, and then you enable IME editing on a window specific window you've got such event. After that all calls to set_ime_position, set_surrounding_text will work, if you don't call enable nothing will work and IME won't be handled at all, and window won't ever popup.

If the client has to call that, they'd still have to permanently call `set_ime_position`, right? I don't see how this changes anything about IME being presented or not. Or does downstream actually have to ask for IME every time? In which case I'd like to know *how* they would know about that?

> It's telling you about string that is currently being edited, but you can't insert it into widget yet. This is a quite common name and concept in IME world tbh. The goal is to display that string to end user, so they know what they're typing, since IME window isn't guaranteed to be presented unlike preedit string, and it makes typing more natural. You generally display that string inline in your text editor and user can move a cursor inside it and edit edit while typing.

So basically the same as `CompositionUpdate`? I don't see how the start/update/end API wouldn't be sufficient then tbh. The `String` payload of start being optional even.

**Comment by kchibisov** [comment]
>If the client has to call that, they'd still have to permanently call set_ime_position, right? I don't see how this changes anything about IME being presented or not. Or does downstream actually have to ask for IME every time? In which case I'd like to know how they would know about that?

It's a `WindowEvent` you don't call for it, you naturally check for it in event handling, and then you may set a state in your app that IME is presented. With `CompositionEnabled` window **event** winit tells you that you **actually** have IME running, on a system without IME you won't ever get that event and so you won't ever call to `set_ime_position` or check for anything other than your apps internal state.

```rust
evlp.run_return(move |event, _, _| {
    match event {
        WindowEvent::CompositionEnabled => {
            inner_state.ime = true;
            window.enable_ime();
        },
        _ => (),
    }

    if inner_state.ime {
        window.set_ime_position(x, y);
    }
});
```
And if you don't call to `enable_ime` it won't ever be enabled meaning that you won't have IME handling for your window. Before you've got `CompositionEnabled` you don't need to call anything from IME stack, since you don't have it. In the current API you must do the following to handle IME which is odd.

```rust
evlp.run_return(move |event, _, _| {
    window.set_ime_position(x, y);
});
```

Meaning you always call to IME even though you don't have IME on your system, which basically results in more work for downstream and for winit.

This was a pseudo code and not real rust.

>So basically the same as CompositionUpdate? I don't see how the start/update/end API wouldn't be sufficient then tbh. The String payload of start being optional even.

Update is the way `Preedit` is called there, you'll be fine with Start/Update/End, but you don't need them, you only need `Preedit` and `Commit` or `Update/End`, I've just named them a bit more IME-ish and removed start, since I don't see a purpose for it.

**Comment by kchibisov** [comment]
Alternative we can embed `enable_ime` in other calls, since user must call `set_ime_position` or `set_surrounding` text for IME to actually work on at least Wayland. We'd still need disable to temporary disable IME though.

**Comment by chrisduerr** [comment]
> It's a WindowEvent you don't call for it, you naturally check for it in event handling, and then you may set a state in your app that IME is presented. With CompositionEnabled window event winit tells you that you actually have IME running, on a system without IME you won't ever get that event and so you won't ever call to set_ime_position or check for anything other than your apps internal state.

Yeah, but doesn't the IME window pop up as soon as `CompositionEnabled` is sent to us? Then the window will appear in the wrong location. Unless you don't show the window until `set_ime_position` is called.

If you have to get an enabled IME event and then explicitly have to ask for things to actually get enabled and then also set the ime position that just seems like an extremely awkward API where nobody would be able to understand what the hell you're supposed to do without looking at an example.

**Comment by kchibisov** [comment]
>Yeah, but doesn't the IME window pop up as soon as CompositionEnabled is sent to us?

No. IME isn't started unless you enable it, that's how at least it work on Wayland. Compositor tells you that you have IME around, and you should react to it with `enable` and then issuing `set_ime_position` or `set_surrounding` text if you want. You react not right away, but after you've got that event. So it's system informs you, you react whether to opt in or not. And also, IME window != IME, since it could be completely missing.

>If you have to get an enabled IME event and then explicitly have to ask for things to actually get enabled and then also set the ime position that just seems like an extremely awkward API where nobody would be able to understand what the hell you're supposed to do without looking at an example.

How are you going to approach it differently? Just calling random IME things all the time isn't an option. We may include enable in set_ime_position and set_surrounding_text, but I don't see how you'll live with the current winit API, since it's nearly impossible to work with IME if you don't have a notion that IME was enabled. The event could be called differently if that's the concern.

**Comment by chrisduerr** [comment]
> We may include enable in set_ime_position and set_surrounding_text

I don't see what the problem would be with that. In general unless a position is set, there's no point in showing it anyways.

> I don't see how you'll live with the current winit API

There's nothing wrong with the current API. It doesn't matter if you check downstream if IME is visible or if winit does it internally. Forcing the downstream to do it just because it makes people feel like it's more efficient is just unnecessary work for everyone. I really don't see how this is "impossible".

**Comment by kchibisov** [comment]
>There's nothing wrong with the current API. It doesn't matter if you check downstream if IME is visible or if winit does it internally. Forcing the downstream to do it just because it makes people feel like it's more efficient is just unnecessary work for everyone. I really don't see how this is "impossible".

Yeah, but then you should compute every draw call ime position, which could be not desired for some applications and slow, issue surrounding text, meaning that you must create strings on each render tick that will be dropped right away, and also set content type, so you will have to call 3 functions constantly.

The point is that the work for clients and winit is unequal, if clients should do the same amount of work as winit, it’ll be fine, but you literally will have 3 functions that you should call.

Also it doesn’t bring more work, since you’re free to ignore those events, however if you handle IME you’ll still have to carry IME state around in clients due to preedit string.

Also on Wayland if you disable IME right in the middle of editing, client must clear preedit string, and without that event winit must fake empty commit or something, which feels hackish.

**Comment by garasubo** [comment]
@runiq No, that PR is not for implementing IME events. We still need to implement new APIs for MacOS.

**Comment by runiq** [comment]
Ah, sorry for the noise then—feel free to hide/delete.

**Comment by garasubo** [comment]
After @kchibisov rebased the branch to the latest master, X11 IME event stopped working somehow. I'm currently investigating why my change stop working with latest master.

**Comment by t18b219k** [comment]
I'm implementing web IME base on egui_web does.
But I have question for IME::Preedit . We must to implement it ?
CompositionEvent Api does not provide Preedit cursor begin and end.

**Comment by komi1230** [comment]
@t18b219k I appreciate your support. About this question, my answer is YES. As you know, enum `IME::Preedit` needs some arguments like `(preedit_text, cursor_start, cursor_end)`. 
In CompositionEvent API in web, there are three event type: `CompostionStart`, `CompositionUpdate` and `CompositionEnd`. I checked the implementation of `web_sys` and found `web_sys` has `CompositionEvent`. Check [this](https://rustwasm.github.io/wasm-bindgen/api/web_sys/struct.CompositionEvent.html). So when you tackle the implementation of web in this issue, you should do in the following step:

1. Include `CompositionEvent` from `web_sys`
2. Implement functions about IME
  a. Get composing string with `composition_event.data().unwrap()`
  b. Fill `cursor_start` and `cursor_end` fields with `composing_string.len()`
  c. Send event to front

Here `cursor_start` and `cursor_end` mean where cursor is, and these data type is expressed in byte-wise.

NOTICE: As I said, `cursor_start` and `cursor_end` are byte-wise and you should NOT count characters.

**Comment by t18b219k** [comment]
Ok I should do same as macOS.

**Comment by t18b219k** [comment]
current status of web 
## Implemented.
- [x] Implement set_ime_position,
- [x] Install HtmlInputElement to take input.
- [x] Install EventHandlers to handle InputEvent,CompositionEvent.
## Bugs
* Backspace not working.
* When composition end fired first char of committed text received .

**Comment by t18b219k** [comment]
Please test #2173 . I dont have macos/ios device so i couldn't test it.
all comment welcome for my work. 
Tested on 
## Firefox
* Firefox 97 windows with ms ime
* Firefox 91.5 esr debian with fcitx/moza
* Firefox 96 ubuntu with fcitx/mozc
* Firefox nightly 99 android with PoBoxPlus and Gboard
## WebkitGTK
* Epiphany 3.38.2 debian
## Blink
* MS edge 98 windows

**Comment by moko256** [comment]
current status of Windows:
## Implemented

PR: #2241

I am testing this with other IME.

**Comment by t18b219k** [comment]
I have question about Web IME on mobile devices.
Shall we open on screen keyboard if set_ime_allowed(false) ?
if don't open onscreen keyboard almost impossible to input text by smart phone or tablets without physical keyboard.
most of on screen keyboards pass input by IME .
which is better ?
 * ignore IME event.
 * open on screen keyboard  but ignore IME event only process Keyboard event.
 * don't open onscreen keyboard.

**Comment by kchibisov** [comment]
I'm pretty sure `set_ime_allowed(false)` should hide keyboard, since you want to hide keyboard in e.g. games and so on.

I'm pretty sure IME input should be processed along side OSK, since that's how it works on Wayland?

You can have keyboard input during IME input, that's fine, it just means that you can receive IME events, handing is up to you?

**Comment by rib** [comment]
Looking a bit into how Winit's IME abstraction can be supported on Android, one thing I've seen is that on Android their IME abstraction is also expected to be able to track a user "selection" which may be separate from the current composition region.

E.g. ref: https://developer.android.com/reference/android/view/inputmethod/InputConnection#cursors,-selections-and-compositions

In particular:
> The composing region and the selection are completely independent of each other, **and the IME may use them however they see fit.**

More practically speaking; one likely way we could look at adding IME support to the Android backend would be via this Game Text Input library from the Android Game Developer Kit:

https://developer.android.com/games/agdk/add-support-for-text-input
And here's the reference for that API:
https://developer.android.com/reference/games/game-text-input/group/game-text-input

In particular that API includes this native [`GameTextInput_setState`](https://developer.android.com/reference/games/game-text-input/group/game-text-input#gametextinput_setstate) function that's used to communicate the state that the Java [`InputConnection`](https://developer.android.com/reference/android/view/inputmethod/InputConnection) wants, which takes a [`GameTextInputState`](https://developer.android.com/reference/games/game-text-input/struct/game-text-input-state#struct_game_text_input_state) struct that can include a selection that's separate from the compose region.

Similarly there is also a [`GameTextInput_getState`](https://developer.android.com/reference/games/game-text-input/group/game-text-input#gametextinput_getstate) and an [event callback](https://developer.android.com/reference/games/game-text-input/group/game-text-input#gametextinputeventcallback) that also use this [`GameTextInputState`](https://developer.android.com/reference/games/game-text-input/struct/game-text-input-state#struct_game_text_input_state)  struct that can report a selection that's separate from the compose region.

I'm wondering if it would be good if Winit's abstraction also allowed for tracking and updating any current user selection, separate from any current compose region to enable Android IMEs to get all the state they want.

Something else I'm wondering is what could be a good way of letting applications distinguish between platforms that move the IME UI via `set_ime_position` to follow wherever input focus is within the application (e.g. IMEs on desktop window systems) vs IMEs like soft keyboards on phones / tablets where it's the other way around and the apps generally need to make sure to update their UI to keep an input field that's the target for an IME within the current insets of the application that may be significantly reduced while the keyboard is visible.

**Comment by Isopod** [comment]
IMEs on Linux still don't work, as far as I can tell. I should say I'm not some kind of IME guru, I only use it for testing. I tested it with Anthy (Japanese IME) on X11. For the record, the IME works in most other programs, which includes Firefox, Chrome and all GTK and Qt applications.

I attempted to enable IME support like this:
```rust
window.set_ime_allowed(true);
```

This makes compose sequences (dead keys, e.g. ´+e = é) work, but not IMEs. In order to support IMEs on Linux, I think winit would need to implement IBus, since this is what most desktop environments (Gnome etc.) use these days.

**Comment by kchibisov** [comment]
Winit supports `XIM` and `text_input_v3`, `dbus` based implementation are out of scope.

GNOME works with both of them out of the box if you enable IME, same with kwin.

Be aware that once you enable ime support, you must handle the `Ime` events.

**Comment by Isopod** [comment]
At the moment I'm just dumping all events to the console. I'm not receiving any IME events. I just get `ReceivedCharacter()` events with latin letters.

**Comment by kchibisov** [comment]
if you use alacritty do you have IME input in it?

**Comment by kchibisov** [comment]
Also, make sure that what you're using is using `XIM`.

**Comment by Isopod** [comment]
I do happen to use alacritty and no, I don't have IME input there, either. When I manually set the input method to XIM by overriding environment variables (e.g.  `GTK_IM_MODULE=xim`), then I don't have any IME input in those applications, either. But that's to be expected, isn't it? After all, I'm using IBus and not XIM. GTK and Qt have their own input modules for IBus, so that's not an issue.

Is IBus really supposed to be completely backward-compatible with XIM? Because that has never been the case for me, the compatibility seems to end at compose sequences/dead keys. I don't know if I'm doing anything wrong. I tried launching ibus with `--xim`, but it doesn't make any difference. I don't use any desktop environment, just i3.

**Comment by kchibisov** [comment]
>Is IBus really supposed to be completely backward-compatible with XIM? Because that has never been the case for me, the compatibility seems to end at compose sequences/dead keys. I don't know if I'm doing anything wrong. I tried launching ibus with --xim, but it doesn't make any difference. I don't use any desktop environment, just i3.

I'd suggest to consult the ibus docs, I know that ibus the way setup on GNOME works without any issue with alacritty on X11 using XIM.


There're plenty IMEs supporting XIM (e.g. fcitx) and surely ibus can do that given that it works on stock fedora gnome for me.

**Comment by notgull** [comment]
> `dbus` based implementation are out of scope.

Do you have a link to discussion on this topic? In my opinion, if IME is in scope then `ibus` should be in scope.

**Comment by kchibisov** [comment]
>Do you have a link to discussion on this topic? In my opinion, if IME is in scope then ibus should be in scope.

Yet on Wayland you don't need it and having ibus where other IMEs exist is not an option (you have uim and so on as well, I never had ibus and use IME daily).

**Comment by Isopod** [comment]
Ok, I just tested it on a fresh Fedora 37 VM, can confirm that the IME works there on both Wayland and X11 (tested in alacritty). The UX isn't the best, but it appears to be usable. Not sure what Fedora is doing differently, it might just be an issue on my machine then.

**Comment by Isopod** [comment]
My issue might be specific to the Arch Linux ibus package. The [PKGBUILD](https://github.com/archlinux/svntogit-community/blob/1002583db43b84fbac4cfd19e47d1611a550caed/trunk/PKGBUILD) does not pass `--enable-xim` to `configure`. I recompiled the package with `--enable-xim` and it works now. Note that you also have to start `ibus-daemon` with `-x` or `--xim`.

**Comment by lucasmerlin** [comment]
I managed to add IME support to android and iOS:
Android input works really well, including autocomplete and suggestions:
<details><summary>Video</summary>
<p>

https://user-images.githubusercontent.com/8009393/251972889-5ac5299f-16d2-429e-899c-6d1e8d31987d.mp4

</p>
</details> 

iOS supports basic text entry, but no autocomplete /autocorrect yet:

<details><summary>Video</summary>
<p>

https://github.com/rust-windowing/winit/assets/8009393/f1836c34-8131-4578-915a-0a0dfbd11ca8

</p>
</details> 

The android support is based on @rib's work on android-activity here: https://github.com/rust-mobile/android-activity/pull/24 
I had to add a new Event to winit, basically just passing through the TextInputState to egui, handling the logic there. 
Implementation in egui is pretty simple, basically 
- send the initial TextInputState (text + cursor position/selection) to winit when focusing a text field + open the keyboard
- update the text in egui whenever a new TextInputState is received, handling the selection and compose region. 

The `TextInputState` struct looks like this:
<details><summary>Expand me</summary>
<p>

```rust

/// This struct holds a span within a region of text from `start` (inclusive) to
/// `end` (exclusive).
///
/// An empty span or cursor position is specified with `Some(start) == Some(end)`.
///
/// An undefined span is specified with start = end = `None`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub struct TextSpan {
    /// The start of the span (inclusive)
    pub start: Option<usize>,

    /// The end of the span (exclusive)
    pub end: Option<usize>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub struct TextInputState {
    pub text: String,
    /// A selection defined on the text.
    pub selection: TextSpan,
    /// A composing region defined on the text.
    pub compose_region: TextSpan,
}
```

</p>
</details> 

For iOS text support, I implemented the UIKeyInput api to receive basic key events but it doesn't support autocomplete / autocorrect / character composition yet. There also is UITextInput to support these features, I think it should be possible to write a implementation in winit that provides the same simple TextInputState api to applications, but I don't have the objective c skills to implement this.

If you want to try these in an egui app, you can add the following to your Cargo.toml to try my branches:
```rust
[patch.crates-io]
winit = { git = "https://github.com/lucasmerlin/winit", branch = "v0.28.x_ime_support" }
egui = { git = "https://github.com/lucasmerlin/egui", branch = "mobile_ime_support"}
eframe = { git = "https://github.com/lucasmerlin/egui", branch = "mobile_ime_support"}
egui-wgpu = { git = "https://github.com/lucasmerlin/egui", branch = "mobile_ime_support"}
android-activity = { git = "https://github.com/lucasmerlin/android-activity", branch = "ime_support"}
```
Or try this example for android: https://github.com/lucasmerlin/rust-android-examples/tree/ime_support_showcase/agdk-eframe

If there is interest to add this to winit I'm happy to open a draft PR.

**Comment by kchibisov** [comment]
Well, we'd need patches upstream to be merged first, you could open a draft PR if you want to though.

**Comment by rib** [comment]
Sorry for the delay following up here @lucasmerlin - very cool that you got something working here.

I'm hoping to get a chance to look at this soon.

It'll be good to compare the egui / winit changes with the ones I experimented with at the time:

https://github.com/rib/winit/tree/android-activity-ime-events
https://github.com/rib/egui/tree/android-winit-ime-support

It'll be good to test this on GameActivity 2.0.2 once I land this PR: https://github.com/rust-mobile/android-activity/pull/88

**Comment by phnaharris** [comment]
> So first of all I don't need CompostionDone event, since this is an event when I'd send things downstream, meaning that I don't need CompositionDeleteSurroundingText. And it seems like I don't need DeleteSurroundingText, since I can model it via CompositionPreedit. CompositionCommit(text: String) is indeed needed so the user will know when to insert text.

Hi @kchibisov, I'm wondering why you said that you don't need `CompositionDeleteSurroundingText` event and how we can model it via `CompositionPreedit`.  Because, the surrounding text is the text around the cursor, even if user committed it to widget or not, as far as I know. I think what you mean is we can modify the preedit, but the problem became a thing if the IME want to delete the surrounding text via text input v3 after the preedit be committed to widget, I think we didn't have a mechanism for doing that yet. 

Now I wanna support this feature about surrounding text. Based on the design of `GTK` for this feature, I need a way for downstream (I'm using `egui`) to send the surrounding text back to `winit` whenever it need, through `Ime::RetrieveSurrounding` event. It would be nice if you have any idea to doing that.

**Comment by kchibisov** [comment]
You should add delete surrounding text, just keep in mind that it doesn't exist without setting surrounding text. The issue is how to communicate that cross platform, but it's a technical detail.

Part of that was done in https://github.com/rust-windowing/winit/pull/2993 . IME will also be reworked for 0.31.

**Comment by phnaharris** [comment]
> it doesn't exist without setting surrounding text

I've taken the work in this pull request and got the idea that how the surrounding text be updated in the event loop. I also implement `set_ime_surrounding_text` for Wayland and call it whenever it changed, especially when text input enabled (and commit it to compositor). But I still haven't receive any `DeleteSurroundingText` event triggered by compositor. Do you have any idea about what I'm missing?

Thank you .

**Comment by kchibisov** [comment]
> But I still haven't receive any DeleteSurroundingText event triggered by compositor. Do you have any idea about what I'm missing?

it's IME dependent, you'll get it only when IME thinks that you should get it. Also, ensure to `commit` when setting surrounding text.

**Comment by phnaharris** [comment]
I understand. But if I reproduce the same situation in the same environment, I should expect the same result right? I've created a text box with gtk and egui (with winit), input the same content and when I go back to change some character, gtk worked fine with surrounding text while winit receive a new `Preedit` event. 

Have you ever got a `DeleteSurroundingText` event in winit before? I just wanna know if the bug come from the flow that I'm implemented was wrong or something else? 

My environment:

- Compositor: Hyprland.
- IME: Fcitx5.
- Toolkit: GTK and egui + winit.

**Comment by kchibisov** [comment]
The event is not handled by winit unless you've added it. Use `WAYLAND_DEBUG=1` to compare patters and use that knowledge to troubleshoot, since usually it's all pretty deterministic.

**Comment by phnaharris** [comment]
Okay, I got the idea. Thank you for your help.

**Comment by kchibisov** [comment]
No point in keeping meta issue.
